system_zero/src/alu/bitops.rs

```diff
@@ -0,0 +1,319 @@
+use plonky2::field::extension_field::Extendable;
+use plonky2::field::field_types::{Field, PrimeField64};
+use plonky2::field::packed_field::PackedField;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+
+use crate::registers::alu::*;
+use crate::registers::NUM_COLUMNS;
+
+/// Interpret the 32 elements of `bits` as bits from low to high of a
+/// u32 and return \sum_i bits[i] 2^i as an element of P.
+fn binary_to_u32<F, P>(bits: [P; 32]) -> P
+where
+    F: Field,
+    P: PackedField<Scalar = F>,
+{
+    bits.into_iter()
+        .enumerate()
+        .map(|(i, b)| b * F::from_canonical_u64(1u64 << i))
+        .sum()
+}
+
+/// As for `binary_to_u32` but uses `builder`.
+fn binary_to_u32_recursively<F, const D: usize>(
+    builder: &mut CircuitBuilder<F, D>,
+    bits: [ExtensionTarget<D>; 32],
+) -> ExtensionTarget<D>
+where
+    F: RichField + Extendable<D>,
+{
+    let terms = bits
```

Comment:
We also have a method `reduce_with_powers_ext_recursive` which could be used. It might use fewer gates since it uses `ReducingExtensionGate`, vs the more general-purpose `ArithmeticExtensionGate`.